File: src/pricing.ts
import type { Session } from "./types";
import { getOpenRouterModel } from "./lib/openrouter-pricing";
import { getCatalogModel } from "./lib/price-catalog";

// ---------------------------------------------------------------------------
// Pricing entry type — used by the Settings Pricing card
// ---------------------------------------------------------------------------

export interface PricingEntry {
  model: string;
  effectiveDate: string; // ISO date
  inputPerMtok: number;
  outputPerMtok: number;
  cacheWrite5mPerMtok: number;
  cacheWrite1hPerMtok: number;
  cacheReadPerMtok: number;
}

/**
 * Anthropic derives every cache rate from the input rate by a fixed
 * multiplier: 5-minute write = 1.25x, 1-hour write = 2x, read = 0.1x.
 * Deriving beats transcribing five numbers per model — transcription is what
 * made the old table wrong (Opus 1h was priced at the 5m rate).
 */
function rate(inputPerMtok: number, outputPerMtok: number): Rate {
  return {
    inputPerMtok,
    outputPerMtok,
    cacheWrite5mPerMtok: inputPerMtok * 1.25,
    cacheWrite1hPerMtok: inputPerMtok * 2,
    cacheReadPerMtok: inputPerMtok * 0.1,
  };
}

function entry(
  model: string,
  effectiveDate: string,
  inputPerMtok: number,
  outputPerMtok: number,
): PricingEntry {
  return { model, effectiveDate, ...rate(inputPerMtok, outputPerMtok) };
}

/** A documented non-Anthropic rate with explicit cache prices. */
function apiEntry(
  model: string,
  effectiveDate: string,
  inputPerMtok: number,
  outputPerMtok: number,
  cacheReadPerMtok: number,
  cacheWritePerMtok: number,
): PricingEntry {
  return {
    model,
    effectiveDate,
    inputPerMtok,
    outputPerMtok,
    cacheWrite5mPerMtok: cacheWritePerMtok,
    // OpenAI documents one cache-write price rather than Anthropic's two TTL
    // tiers, so both stored fields use that documented price.
    cacheWrite1hPerMtok: cacheWritePerMtok,
    cacheReadPerMtok,
  };
}

// ---------------------------------------------------------------------------
// Pricing table — current rates for models observed in supported transcripts.
// Rates are per million tokens ($/Mtok), from their first-party price sheets.
// `effectiveDate` is the model release date where known, otherwise the date
// its current official rate was verified; it is shown in Settings but is not
// yet used for historical lookup (see the note on rateFor below).
// ---------------------------------------------------------------------------

const OBSERVED_NON_ANTHROPIC_PRICING: PricingEntry[] = [
  // Verified 2026-09-17 against official provider pricing. See
  // notes/research/2026-09-17-observed-model-pricing.md for sources/caveats.
  apiEntry("gpt-5.5", "2026-09-17", 5, 30, 0.5, 0),
  apiEntry("gpt-5.6-luna", "2026-09-17", 0.2, 1.2, 0.02, 0.25),
  apiEntry("gpt-5.6-terra", "2026-09-17", 2, 12, 0.2, 2.5),
  apiEntry("gpt-5.6-sol", "2026-09-17", 4, 20, 0.4, 5),
  apiEntry("gpt-6-astra", "2026-09-17", 10, 50, 1, 12.5),
  // Official Codex/Work alias for GPT-5.4, not a standalone API SKU.
  apiEntry("codex-auto-review", "2026-09-17", 2.5, 15, 0.25, 0),
  // This bare family ID is free only for the Gemini API/local interpretation;
  // do not generalise this to a paid third-party Gemma endpoint.
  apiEntry("gemma-4", "2026-09-17", 0, 0, 0, 0),
];

export const PRICING_TABLE: PricingEntry[] = [
  // --- Fable / Mythos tier ---
  entry("claude-fable-5", "2026-05-19", 10, 50),
  // --- Opus: current tier ($5/$25 since Opus 4.5) ---
  entry("claude-opus-5", "2026-06-24", 5, 25),
  entry("claude-opus-4-8", "2026-04-22", 5, 25),
  entry("claude-opus-4-7", "2026-02-10", 5, 25),
  entry("claude-opus-4-6", "2025-12-16", 5, 25),
  entry("claude-opus-4-5", "2025-11-24", 5, 25),
  // --- Opus: legacy tier ($15/$75) ---
  entry("claude-opus-4-1", "2025-08-05", 15, 75),
  entry("claude-opus-4", "2025-05-22", 15, 75),
  entry("claude-3-opus", "2024-03-04", 15, 75),
  // --- Sonnet (every generation has shipped at $3/$15) ---
  entry("claude-sonnet-5", "2026-05-05", 3, 15),
  entry("claude-sonnet-4-6", "2025-11-24", 3, 15),
  entry("claude-sonnet-4-5", "2025-09-29", 3, 15),
  entry("claude-3-7-sonnet", "2025-02-24", 3, 15),
  entry("claude-3-5-sonnet", "2024-06-20", 3, 15),
  // --- Haiku ---
  entry("claude-haiku-4-5", "2025-10-15", 1, 5),
  entry("claude-3-5-haiku", "2024-11-05", 0.8, 4),
  entry("claude-3-haiku", "2024-03-07", 0.25, 1.25),
  ...OBSERVED_NON_ANTHROPIC_PRICING,
];

// ---------------------------------------------------------------------------
// Rate lookup
//
// Family-keyword matching, mirroring crates/usage-core/src/pricing.rs. Model
// ids arrive in several shapes — "claude-opus-5", "claude-opus-4-8", and
// reversed vendor forms like "claude-5-sonnet-anthropic" — so we match on the
// family keyword rather than on exact slugs.
//
// Exact, first-party documented non-Anthropic IDs come first, then Anthropic
// family matching. Anything else falls back to the downloaded price catalog
// (lib/price-catalog.ts), then the bundled OpenRouter snapshot, then remains
// "rate unknown"; charging an unknown model at a guessed rate produces
// confident wrong numbers, which is worse.
//
// ponytail: rates are current-only, not effective-dated. Known gap: Claude
// Sonnet 5 bills at an introductory $2/$10 through 2026-08-31, so Sonnet 5
// spend before that date reads ~33% high. Add date-effective lookup here and
// in pricing.rs when the introductory windows matter.
// ---------------------------------------------------------------------------

export type Rate = Omit<PricingEntry, "model" | "effectiveDate">;

const DOCUMENTED_NON_ANTHROPIC_RATES = new Map(
  OBSERVED_NON_ANTHROPIC_PRICING.map(({ model, effectiveDate: _effectiveDate, ...rate }) => [model, rate]),
);

/** Models that predate the Opus price drop and still bill at $15/$75. */
const LEGACY_OPUS = [
  "claude-3-opus",
  "claude-opus-4-0",
  "claude-opus-4-1",
  "claude-opus-4.1",
  "opus-4-1",
];

export function rateFor(model: string | null): Rate | null {
  const raw = (model ?? "").toLowerCase();
  const m = raw.replace(/:batch$/, "").replace(/:free$/, "");

  const documentedRate = DOCUMENTED_NON_ANTHROPIC_RATES.get(m);
  if (documentedRate) return documentedRate;

  // Placeholder Claude Code uses for locally generated messages; always zero
  // tokens, never billed.
  if (m === "<synthetic>") return null;

  if (m.includes("fable") || m.includes("mythos")) return rate(10, 50);

  if (m.includes("opus")) {
    const legacy = LEGACY_OPUS.some((slug) => m.includes(slug)) || m === "claude-opus-4";
    return legacy ? rate(15, 75) : rate(5, 25);
  }

  if (m.includes("haiku")) {
    if (m.includes("claude-3-haiku")) return rate(0.25, 1.25);
    if (m.includes("3-5-haiku") || m.includes("3.5-haiku")) return rate(0.8, 4);
    return rate(1, 5); // Haiku 4.5 and later
  }

  // Must come after haiku: "claude-3-5-haiku" also contains "claude-3-5".
  if (m.includes("sonnet") || m.includes("claude-3-5") || m.includes("claude-3-7")) {
    return rate(3, 15);
  }

  return openRouterRate(m);
}

/**
 * Fallback rates for non-Anthropic models: the downloaded catalog first, then
 * the snapshot bundled with the app. Both are OpenRouter's list, keyed by its
 * namespaced ids.
 *
 * Codex logs a bare model id ("gpt-5.6-terra"); OpenRouter namespaces it
 * ("openai/gpt-5.6-terra"). Exact, official rates for observed Codex models
 * take priority in `rateFor`; anything that matches none remains unknown.
 */
function openRouterRate(model: string): Rate | null {
  const downloaded = getCatalogModel(model) ?? getCatalogModel(`openai/${model}`);
  if (downloaded) {
    return {
      inputPerMtok: downloaded.inputPerMtok,
      outputPerMtok: downloaded.outputPerMtok,
      cacheWrite5mPerMtok: downloaded.cacheWritePerMtok,
      cacheWrite1hPerMtok: downloaded.cacheWrite1hPerMtok || downloaded.cacheWritePerMtok,
      cacheReadPerMtok: downloaded.cacheReadPerMtok,
    };
  }

  const entry = getOpenRouterModel(model) ?? getOpenRouterModel(`openai/${model}`);
  if (!entry) return null;
  // OpenRouter uses negative sentinel values (e.g. -1_000_000) for router
  // models like "openrouter/auto" where pricing is dynamic and unknown at
  // query time. Treat any negative input rate as unpriced rather than
  // producing a nonsensical negative cost.
  if (entry.promptPerMtok < 0 || entry.completionPerMtok < 0) return null;
  return {
    inputPerMtok: entry.promptPerMtok,
    outputPerMtok: entry.completionPerMtok,
    cacheWrite5mPerMtok: entry.cacheWritePerMtok,
    // OpenAI has no 1-hour cache tier; fall back to the single write rate so
    // the session-level estimate (which charges cache writes at 1h) doesn't
    // silently bill them at zero.
    cacheWrite1hPerMtok: entry.cacheWrite1hPerMtok || entry.cacheWritePerMtok,
    cacheReadPerMtok: entry.cacheReadPerMtok,
  };
}

/** True when we have a real rate for this model. */
export function isModelPriced(model: string | null): boolean {
  if (!model) return true; // null model → nothing to warn about
  return rateFor(model) !== null;
}

/**
 * Estimate the cost in USD for a single API request.
 *
 * Cache writes must be passed split by TTL tier — the two rates differ by
 * 1.6x and real transcripts are a ~22/78 mix, so collapsing them is not a
 * rounding error. Returns null when the model has no known rate.
 */
export function estimateRequestCost(
  model: string | null,
  inputTokens: number,
  outputTokens: number,
  cacheWrite5m: number,
  cacheWrite1h: number,
  cacheRead: number,
): number | null {
  const r = rateFor(model);
  if (!r) return null;
  const M = 1_000_000;
  return (
    (inputTokens / M) * r.inputPerMtok +
    (outputTokens / M) * r.outputPerMtok +
    (cacheWrite5m / M) * r.cacheWrite5mPerMtok +
    (cacheWrite1h / M) * r.cacheWrite1hPerMtok +
    (cacheRead / M) * r.cacheReadPerMtok
  );
}

// ---------------------------------------------------------------------------
// Context windows
//
// Only the models below ship a 1M-token window; everything else — including
// Haiku 4.5 and every Claude 3.x / 4.0-4.5 model — is 200K. The default is
// deliberately the smaller one: under-stating the ceiling shows a fuller bar,
// and a false context alarm is a better failure than a missed one.
// ---------------------------------------------------------------------------

const LONG_CONTEXT_MODELS = [
  "fable",
  "mythos",
  "opus-5",
  "opus-4-8",
  "opus-4.8",
  "opus-4-7",
  "opus-4.7",
  "opus-4-6",
  "opus-4.6",
  "sonnet-5",
  "sonnet-4-6",
  "sonnet-4.6",
];

export const DEFAULT_CONTEXT_WINDOW = 200_000;
export const LONG_CONTEXT_WINDOW = 1_000_000;
/**
 * What Codex reports as `model_context_window` in every rollout it writes.
 * ponytail: hardcoded because the window is per-request data the backend
 * doesn't keep — read it off the session if a second value ever shows up.
 * Without it, gpt sessions measure against 200K and report >100% full.
 */
export const CODEX_CONTEXT_WINDOW = 258_400;

/** Context-window size in tokens for a model id. */
export function contextWindow(model: string | null): number {
  const m = (model ?? "").toLowerCase();
  if (LONG_CONTEXT_MODELS.some((slug) => m.includes(slug))) return LONG_CONTEXT_WINDOW;
  if (m.startsWith("gpt-") || m.startsWith("openai/")) return CODEX_CONTEXT_WINDOW;
  return DEFAULT_CONTEXT_WINDOW;
}

// ---------------------------------------------------------------------------
// Model color mapping
// Returns the CSS variable reference (e.g. "var(--series-1)") for a model.
// Series 1–12 are assigned by model family so the same family always gets the
// same color across all charts.
// ---------------------------------------------------------------------------

// The series tokens in tokens.css name the model each colour belongs to; this
// map must agree with them. It previously stopped at the 4.x generation, so
// claude-sonnet-5, claude-fable-5 and every unknown model fell through to the
// series-1 fallback — the same blue as Opus, which is why a mixed-model chart
// came out one colour. Specific generations are matched before family.
const MODEL_COLOR_MAP: Array<{ pattern: RegExp; series: number }> = [
  { pattern: /gpt-6|astra/i,                     series: 8 },
  { pattern: /gpt-5[.]6-sol/i,                   series: 9 },
  { pattern: /gpt-5[.]6-terra/i,                 series: 10 },
  { pattern: /gpt-5[.]6-luna/i,                  series: 11 },
  { pattern: /gpt-5[.]5/i,                       series: 12 },
  { pattern: /fable|mythos/i,                    series: 3 },
  { pattern: /opus-5|opus-4[-.]6|opus-4[-.]7/i,  series: 1 },
  { pattern: /opus-4[-.]8/i,                     series: 4 },
  { pattern: /opus/i,                            series: 1 },
  { pattern: /sonnet-4[-.]6|4[-.]6-sonnet/i,     series: 6 },
  { pattern: /sonnet-5|claude-5-sonnet/i,        series: 2 },
  { pattern: /claude-3-5-sonnet|sonnet.*3[-.]5/i, series: 7 },
  { pattern: /claude-3-7-sonnet|sonnet.*3[-.]7/i, series: 4 },
  { pattern: /sonnet/i,                          series: 2 },
  { pattern: /haiku-4|4[-.]5-haiku|claude-haiku-4/i, series: 5 },
  { pattern: /haiku/i,                           series: 5 },
];

export function modelColor(model: string): string {
  for (const { pattern, series } of MODEL_COLOR_MAP) {
    if (pattern.test(model)) {
      return `var(--series-${series})`;
    }
  }
  // Keep unrecognised models categorical too. A stable hash preserves the
  // colour between views without collapsing every new model into one grey.
  let hash = 2166136261;
  for (let i = 0; i < model.length; i++) {
    hash ^= model.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  const fallbackSeries = [8, 9, 10, 11, 12];
  return `var(--series-${fallbackSeries[(hash >>> 0) % fallbackSeries.length]})`;
}

// ---------------------------------------------------------------------------
// Session-level helpers
// ---------------------------------------------------------------------------

export function totalTokens(s: Session): number {
  return s.input_tokens + s.output_tokens + s.cache_creation_tokens + s.cache_read_tokens;
}

/**
 * Returns the cost in USD for a session.
 *
 * Prefers `cost_usd` when it is > 0 (set by the Rust backend from per-request
 * accumulation). Falls back to a single-rate front-end estimate for sessions
 * where the backend hasn't populated the field (e.g. legacy data, non-Claude
 * sources). The fallback uses the 1-hour cache-write tier for
 * `cache_creation_tokens` to stay consistent with the backend.
 */
export function estimatedCostUsd(s: Session): number {
  if (s.cost_usd > 0) {
    return s.cost_usd;
  }
  // Fallback for sources that don't compute per-request cost. The 5m/1h split
  // isn't retained at session level, so all cache creation is charged at the
  // 1h tier. Unpriced models contribute 0 — `isModelPriced` drives the
  // "cost estimates may be inaccurate" warning so a 0 is never silent.
  return (
    estimateRequestCost(
      s.model,
      s.input_tokens,
      s.output_tokens,
      0,
      s.cache_creation_tokens,
      s.cache_read_tokens,
    ) ?? 0
  );
}

File: crates/usage-core/src/pricing.rs
use crate::model::Session;

/// Per-token rates for a single model, in USD per million tokens.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Rate {
    pub input: f64,
    pub output: f64,
    /// Cache write at the 5-minute TTL tier.
    pub cache_write_5m: f64,
    /// Cache write at the 1-hour TTL tier.
    pub cache_write_1h: f64,
    pub cache_read: f64,
}

impl Rate {
    /// Anthropic derives every cache rate from the input rate by a fixed
    /// multiplier: 5-minute write = 1.25x, 1-hour write = 2x, read = 0.1x.
    /// Deriving them beats transcribing five numbers per model — transcription
    /// is what made the old table wrong (opus 1h was priced at the 5m rate).
    fn anthropic(input: f64, output: f64) -> Self {
        Self {
            input,
            output,
            cache_write_5m: input * 1.25,
            cache_write_1h: input * 2.0,
            cache_read: input * 0.1,
        }
    }

    /// Providers that publish a single cache-write rate rather than separate
    /// 5-minute and 1-hour TTL tiers use that rate for both stored fields.
    fn api(input: f64, output: f64, cache_write: f64, cache_read: f64) -> Self {
        Self {
            input,
            output,
            cache_write_5m: cache_write,
            cache_write_1h: cache_write,
            cache_read,
        }
    }
}

// ---------------------------------------------------------------------------
// Rate table — Anthropic families plus exact, documented Codex model IDs.
//
// Exact rates for model IDs observed in Codex transcripts are listed below.
// Everything else returns None and is reported as "rate unknown" rather than
// being charged at a guessed rate — the previous table billed
// `nvidia-nemotron-nano-3-30b` at Sonnet rates, which is worse than admitting
// we don't know.
//
// The frontend also has an OpenRouter-derived fallback for non-observed IDs.
// Keep the exact entries in both layers aligned so backend scans and frontend
// fallback calculations agree.
// ---------------------------------------------------------------------------

/// Models that predate the price drop and still bill at the legacy Opus rate.
/// Closed set — everything newer is on the $5/$25 tier, so unknown Opus models
/// default to the current tier rather than the legacy one.
const LEGACY_OPUS: [&str; 5] = [
    "claude-3-opus",
    "claude-opus-4-0",
    "claude-opus-4-1",
    "claude-opus-4.1",
    "opus-4-1",
];

pub fn rate_for(model: &str) -> Option<Rate> {
    let m = model.to_lowercase();
    let m = m.trim_end_matches(":batch").trim_end_matches(":free");

    // Placeholder used by Claude Code for locally generated messages; always
    // zero tokens, never billed.
    if m == "<synthetic>" {
        return None;
    }

    // Exact IDs observed in Codex transcripts, verified against their
    // providers' official price sheets on 2026-09-17. `codex-auto-review` is
    // a rate-card alias for GPT-5.4; bare `gemma-4` is zero only for its
    // documented Gemini API Free Tier/local interpretation. See
    // notes/research/2026-09-17-observed-model-pricing.md for sources.
    let documented = match m {
        "gpt-5.5" => Some(Rate::api(5.0, 30.0, 0.0, 0.5)),
        "gpt-5.6-luna" => Some(Rate::api(0.2, 1.2, 0.25, 0.02)),
        "gpt-5.6-terra" => Some(Rate::api(2.0, 12.0, 2.5, 0.2)),
        "gpt-5.6-sol" => Some(Rate::api(4.0, 20.0, 5.0, 0.4)),
        "gpt-6-astra" => Some(Rate::api(10.0, 50.0, 12.5, 1.0)),
        "codex-auto-review" => Some(Rate::api(2.5, 15.0, 0.0, 0.25)),
        "gemma-4" => Some(Rate::api(0.0, 0.0, 0.0, 0.0)),
        _ => None,
    };
    if documented.is_some() {
        return documented;
    }

    // Model ids appear in several shapes: "claude-opus-5", "claude-opus-4-8",
    // and reversed vendor forms like "claude-5-sonnet-anthropic". Family
    // keyword matching handles all of them.

    // Newer generations break the 0.1x cache-read rule, so they're matched by
    // version before their family. Verified against Anthropic's published
    // rates on 2026-09-23. Mythos 5.1 stays on the family rate: whether it
    // shares Fable 5.1's cache-read price is unannounced.
    if m.contains("fable-5-1") || m.contains("fable-5.1") {
        return Some(Rate { cache_read: 0.25, ..Rate::anthropic(10.0, 50.0) });
    }
    if m.contains("opus-5-5") || m.contains("opus-5.5") {
        return Some(Rate { cache_read: 0.2, ..Rate::anthropic(4.0, 20.0) });
    }

    if m.contains("fable") || m.contains("mythos") {
        return Some(Rate::anthropic(10.0, 50.0));
    }

    if m.contains("opus") {
        if LEGACY_OPUS.iter().any(|legacy| m.contains(legacy)) || m == "claude-opus-4" {
            return Some(Rate::anthropic(15.0, 75.0));
        }
        return Some(Rate::anthropic(5.0, 25.0));
    }

    if m.contains("haiku") {
        if m.contains("claude-3-haiku") || m.ends_with("/claude-3-haiku") {
            return Some(Rate::anthropic(0.25, 1.25));
        }
        if m.contains("3-5-haiku") || m.contains("3.5-haiku") {
            return Some(Rate::anthropic(0.8, 4.0));
        }
        // Haiku 4.5 and later.
        return Some(Rate::anthropic(1.0, 5.0));
    }

    if m.contains("sonnet") || m.contains("claude-3-5") || m.contains("claude-3-7") {
        // Every Sonnet from 3.5 through 5 has shipped at $3/$15. Must come
        // after haiku: "claude-3-5-haiku" also contains "claude-3-5".
        return Some(Rate::anthropic(3.0, 15.0));
    }

    None
}

/// Estimated cost in USD for a single API request.
///
/// Returns `None` when the model has no known rate — callers should surface
/// that as "unknown", not as $0.
pub fn estimate_request_cost(
    model: &str,
    input: u64,
    output: u64,
    cache_write_5m: u64,
    cache_write_1h: u64,
    cache_read: u64,
) -> Option<f64> {
    let rate = rate_for(model)?;
    let m = 1_000_000.0;
    Some(
        input as f64 / m * rate.input
            + output as f64 / m * rate.output
            + cache_write_5m as f64 / m * rate.cache_write_5m
            + cache_write_1h as f64 / m * rate.cache_write_1h
            + cache_read as f64 / m * rate.cache_read,
    )
}

/// Estimated cost in USD for a Session.
///
/// When `cost_usd` is already populated (set by per-request accumulation
/// during scanning), use it directly. Otherwise fall back to a single-rate
/// estimate from the session's aggregate tokens — useful for sessions from
/// sources that don't support per-request parsing.
pub fn estimate_cost(session: &Session) -> f64 {
    if session.cost_usd > 0.0 {
        return session.cost_usd;
    }
    // Fallback: single-rate estimate. The 5m/1h split isn't retained at
    // session level, so all cache creation is charged at the 1h tier.
    estimate_request_cost(
        session.model.as_deref().unwrap_or(""),
        session.input_tokens,
        session.output_tokens,
        0,
        session.cache_creation_tokens,
        session.cache_read_tokens,
    )
    .unwrap_or(0.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The bug this table was rewritten to fix: `contains("opus")` used to
    /// return Claude 3 Opus rates for every Opus model, overcharging Opus 5
    /// and 4.8 by ~170%.
    #[test]
    fn current_opus_is_not_billed_at_legacy_opus_rates() {
        let opus5 = rate_for("claude-opus-5").unwrap();
        assert_eq!(opus5.input, 5.0);
        assert_eq!(opus5.output, 25.0);
        assert_eq!(rate_for("claude-opus-4-8").unwrap(), opus5);

        let legacy = rate_for("claude-3-opus").unwrap();
        assert_eq!(legacy.input, 15.0);
        assert_eq!(rate_for("claude-opus-4-1").unwrap(), legacy);
    }

    #[test]
    fn cache_rates_are_derived_from_input() {
        let r = rate_for("claude-sonnet-5").unwrap();
        assert_eq!(r.input, 3.0);
        assert_eq!(r.cache_write_5m, 3.75);
        assert_eq!(r.cache_write_1h, 6.0);
        assert!((r.cache_read - 0.3).abs() < 1e-12);
    }

    #[test]
    fn families_resolve_across_id_shapes() {
        assert_eq!(rate_for("claude-fable-5").unwrap().input, 10.0);
        // Reversed vendor form seen in real transcripts.
        assert_eq!(rate_for("claude-5-sonnet-anthropic").unwrap().input, 3.0);
        assert_eq!(rate_for("claude-4.5-haiku-anthropic").unwrap().input, 1.0);
        assert_eq!(rate_for("claude-haiku-4-5-20251001").unwrap().input, 1.0);
        assert_eq!(rate_for("claude-3-5-haiku").unwrap().input, 0.8);
        assert_eq!(rate_for("claude-3-haiku").unwrap().input, 0.25);
    }

    #[test]
    fn observed_codex_models_have_documented_rates() {
        let sol = rate_for("gpt-5.6-sol").unwrap();
        assert_eq!(sol.input, 4.0);
        assert_eq!(sol.output, 20.0);
        assert_eq!(sol.cache_write_5m, 5.0);
        assert_eq!(sol.cache_read, 0.4);

        let astra = rate_for("gpt-6-astra").unwrap();
        assert_eq!(astra.input, 10.0);
        assert_eq!(astra.output, 50.0);
        assert_eq!(astra.cache_write_1h, 12.5);
        assert_eq!(astra.cache_read, 1.0);

        let auto_review = rate_for("codex-auto-review").unwrap();
        assert_eq!(auto_review.input, 2.5);
        assert_eq!(auto_review.output, 15.0);
        assert_eq!(auto_review.cache_read, 0.25);

        assert_eq!(rate_for("gemma-4").unwrap(), Rate {
            input: 0.0,
            output: 0.0,
            cache_write_5m: 0.0,
            cache_write_1h: 0.0,
            cache_read: 0.0,
        });
    }

    #[test]
    fn newest_generations_use_their_own_cache_read_rate() {
        for id in ["claude-opus-5-5", "claude-opus-5.5", "anthropic/claude-opus-5.5"] {
            let r = rate_for(id).unwrap();
            assert_eq!((r.input, r.output, r.cache_read), (4.0, 20.0, 0.2), "{id}");
            assert_eq!((r.cache_write_5m, r.cache_write_1h), (5.0, 8.0), "{id}");
        }
        for id in ["claude-fable-5-1", "anthropic/claude-fable-5.1"] {
            let r = rate_for(id).unwrap();
            assert_eq!((r.input, r.output, r.cache_read), (10.0, 50.0, 0.25), "{id}");
        }
        // The previous generation is unchanged.
        assert_eq!(rate_for("claude-opus-5").unwrap().cache_read, 0.5);
        assert_eq!(rate_for("claude-fable-5").unwrap().cache_read, 1.0);
    }

    #[test]
    fn unknown_models_are_unpriced_rather_than_guessed() {
        for model in [
            "nvidia-nemotron-nano-3-30b-aws",
            "openrouter/auto",
            "z-ai/glm-5.2",
            "openai.gpt-5.6-terra",
            "<synthetic>",
            "",
        ] {
            assert!(rate_for(model).is_none(), "{model} should be unpriced");
            assert!(estimate_request_cost(model, 100, 100, 100, 100, 100).is_none());
        }
    }

    #[test]
    fn request_cost_charges_each_tier_at_its_own_rate() {
        // sonnet-5: in 3, out 15, cw5m 3.75, cw1h 6, cr 0.3
        let cost = estimate_request_cost("claude-sonnet-5", 1000, 1000, 1000, 1000, 1000).unwrap();
        let expected = (3.0 + 15.0 + 3.75 + 6.0 + 0.3) / 1000.0;
        assert!((cost - expected).abs() < 1e-12, "{cost} != {expected}");
    }
}
